servers/audio_server.cpp

```diff
@@ -244,7 +248,46 @@ AudioDriver *AudioDriverManager::get_driver(int p_driver) {
 //////////////////////////////////////////////
 //////////////////////////////////////////////
 
-void AudioServer::_driver_process(int p_frames, int32_t *p_buffer) {
+AudioServer *AudioServer::singleton = nullptr;
+
+void AudioServer::_write_sample(void *p_buffer, int p_idx, float p_sample) {
+	if (unlikely(p_buffer == nullptr))
+		return;
+
+	switch (AudioDriver::get_singleton()->get_output_buffer_format()) {
+		case AudioDriver::BUFFER_FORMAT_INTEGER_8:
+			((int8_t *)p_buffer)[p_idx] = p_sample * ((1u << (8 - 1)) - 1);
+			break;
+
+		case AudioDriver::BUFFER_FORMAT_INTEGER_16:
+			((int16_t *)p_buffer)[p_idx] = p_sample * ((1u << (16 - 1)) - 1);
+			break;
+
+		case AudioDriver::BUFFER_FORMAT_INTEGER_24: {
+			int32_t sample = p_sample * ((1u << (24 - 1)) - 1);
+			int32_t sign = p_sample < 0 ? -1 : 1;
+
+			((int8_t *)p_buffer)[p_idx * 3 + 2] = sign * (ABS(sample) >> 16);
+			((int8_t *)p_buffer)[p_idx * 3 + 1] = sign * (ABS(sample) >> 8);
+			((int8_t *)p_buffer)[p_idx * 3 + 0] = sample;
+		} break;
+
+		case AudioDriver::BUFFER_FORMAT_INTEGER_32:
+			// (1u << (32 - 1)) - 1 can't be stored in float correctly.
+			((int32_t *)p_buffer)[p_idx] = (double)p_sample * ((1u << (32 - 1)) - 1);
```

Comment:
It is different from the implementation that was before:
https://github.com/godotengine/godot/blob/9bc49a66bae5e9e506f12df3b3e141c8da13f983/servers/audio_server.cpp#L288-L291
But I'm not sure why it was like this, all I can find on the Internet is how to convert from float to 16-bit integer and it's just to multiply a float to match the range of the number (eg. -1.0 is the minimum value that can fit in 16-bit integer and 1.0 is the maximum), so I expect it to be the same with 32-bit integers?
Also, used everywhere `((1u << (*bits* - 1)) - 1)`, though it would be better to use `(1u << (*bits* - 1))` for negative numbers and `((1u << (*bits* - 1)) - 1)` for positive ones. But I think it's really hard to hear the difference so kept it like this for a small optimization (which would be good as audio callbacks really should run as fast as possible).